In [2]:
print ("Привет")


Привет


## Задание 1.
1. Скачать данные по ссылке
https://drive.google.com/file/d/1MpAdHAl727fO3oW32NO4FpSRhU
BUfjfS/edit
2. Считать данные с помощью pandas
3. Вывести на экран первые 5 строк
1.1 Создать новый признак Cpu_Company, который будет
содержать только название фирмы, которая произвела CPU
1.2 Создать новый признак Memory_Amount, который будет
содержать только количество Gb памяти без указания типа
носителя
1.3 Создать новый признак Memory_Type, который будет
содержать только тип носителя (HDD/SDD/др.)
1.4 Удалите признаки Memory и ScreenResolution

In [3]:
import pandas as pd

In [4]:
laptop = pd.read_csv('laptop_price.csv', encoding ='latin-1')
laptop.head(5)

NameError: name 'pd' is not defined

In [5]:
#laptop['Cpu_Company'] = laptop['Cpu'].apply(lambda i:i.split(' ')[0]) # вытащили первое слово из столбца в новый столбец
laptop['Cpu_Company'] = laptop['Cpu'].str.split().str.get(0) #второй вариант решения
laptop.head()

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   
3          4   Apple  MacBook Pro  Ultrabook    15.4   
4          5   Apple  MacBook Pro  Ultrabook    13.3   

                     ScreenResolution                         Cpu   Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

                Memory                           Gpu  OpSys  Weight  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS  1.37kg   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS  1.34kg   
2            256GB SSD         Intel HD Graphics 620  No OS  1.86kg   
3            512GB SSD            AMD Radeon Pro 455  macOS  1.83kg   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS  1.37kg   

   Price_euros Cpu_Company  
0      1339.69       Intel  
1       898.94       Intel  
2       575.00       Intel  
3      2537.45       Intel  
4      1803.60       Intel

In [6]:
laptop['Memory_Amount'] = laptop['Memory'].apply(lambda i:i.split(' ')[0])
laptop.head()

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   
3          4   Apple  MacBook Pro  Ultrabook    15.4   
4          5   Apple  MacBook Pro  Ultrabook    13.3   

                     ScreenResolution                         Cpu   Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

                Memory                           Gpu  OpSys  Weight  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS  1.37kg   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS  1.34kg   
2            256GB SSD         Intel HD Graphics 620  No OS  1.86kg   
3            512GB SSD            AMD Radeon Pro 455  macOS  1.83kg   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS  1.37kg   

   Price_euros Cpu_Company Memory_Amount  
0      1339.69       Intel         128GB  
1       898.94       Intel         128GB  
2       575.00       Intel         256GB  
3      2537.45       Intel         512GB  
4      1803.60       Intel         256GB

In [7]:
laptop['Memory_Type'] = laptop['Memory'].apply(lambda i:i.split(' ')[1])
laptop.head()

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   
3          4   Apple  MacBook Pro  Ultrabook    15.4   
4          5   Apple  MacBook Pro  Ultrabook    13.3   

                     ScreenResolution                         Cpu   Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

                Memory                           Gpu  OpSys  Weight  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS  1.37kg   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS  1.34kg   
2            256GB SSD         Intel HD Graphics 620  No OS  1.86kg   
3            512GB SSD            AMD Radeon Pro 455  macOS  1.83kg   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS  1.37kg   

   Price_euros Cpu_Company Memory_Amount Memory_Type  
0      1339.69       Intel         128GB         SSD  
1       898.94       Intel         128GB       Flash  
2       575.00       Intel         256GB         SSD  
3      2537.45       Intel         512GB         SSD  
4      1803.60       Intel         256GB         SSD

Что делает строка целиком:
Она извлекает числовое значение количества памяти из строкового описания и создаёт новый столбец Memory_Amount.

Пошаговая расшифровка:
1. laptop['Memory']
Обращение к столбцу Memory в DataFrame laptop

Пример данных: ['8 GB', '16 GB DDR4', '512 GB SSD']

2. .apply(lambda i: ...)
Применяет функцию к КАЖДОЙ ячейке столбца Memory

lambda i: — анонимная функция, где i это одна ячейка

3. i.split(' ')
Разбивает строку по пробелам

Пример: '16 GB DDR4'.split(' ') → ['16', 'GB', 'DDR4']

4. [0]
Берёт первый элемент из получившегося списка

Пример: ['16', 'GB', 'DDR4'][0] → '16'

5. laptop['Memory_Amount'] = ...
Создаёт новый столбец Memory_Amount с результатами

In [8]:
def convert_to_gb(x):
    memory=x.split(' ')[0]
    if memory.endswith('GB'):
        return int(memory.replace('GB' , ''))
    else:
        return int(float(memory.replace('TB' , ''))*1024)
laptop['Memory_Amount1'] = laptop['Memory'].apply(convert_to_gb)
# laptop['Memory_Amount'] = laptop['Memory'].apply(lambda i:''.join(filter(str.isdigit, i))) #без перевода в гиги
laptop['Memory_Amount2'] = laptop['Memory'].apply(lambda i:int(i.split(' ')[0].replace('GB', '')) if 'GB' in i else int(float(i.split(' ')[0].replace('TB', '')) * 1024))
laptop.head(20)

laptop_ID Company            Product   TypeName  Inches  \
0           1   Apple        MacBook Pro  Ultrabook    13.3   
1           2   Apple        Macbook Air  Ultrabook    13.3   
2           3      HP             250 G6   Notebook    15.6   
3           4   Apple        MacBook Pro  Ultrabook    15.4   
4           5   Apple        MacBook Pro  Ultrabook    13.3   
5           6    Acer           Aspire 3   Notebook    15.6   
6           7   Apple        MacBook Pro  Ultrabook    15.4   
7           8   Apple        Macbook Air  Ultrabook    13.3   
8           9    Asus    ZenBook UX430UN  Ultrabook    14.0   
9          10    Acer            Swift 3  Ultrabook    14.0   
10         11      HP             250 G6   Notebook    15.6   
11         12      HP             250 G6   Notebook    15.6   
12         13   Apple        MacBook Pro  Ultrabook    15.4   
13         14    Dell      Inspiron 3567   Notebook    15.6   
14         15   Apple        MacBook 12"  Ultrabook    12.0   
15         16   Apple        MacBook Pro  Ultrabook    13.3   
16         17    Dell      Inspiron 3567   Notebook    15.6   
17         18   Apple        MacBook Pro  Ultrabook    15.4   
18         19  Lenovo  IdeaPad 320-15IKB   Notebook    15.6   
19         20    Dell             XPS 13  Ultrabook    13.3   

                             ScreenResolution                         Cpu  \
0          IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   
1                                    1440x900        Intel Core i5 1.8GHz   
2                           Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   
3          IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz   
4          IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   
5                                    1366x768     AMD A9-Series 9420 3GHz   
6          IPS Panel Retina Display 2880x1800        Intel Core i7 2.2GHz   
7                                    1440x900        Intel Core i5 1.8GHz   
8                           Full HD 1920x1080  Intel Core i7 8550U 1.8GHz   
9                 IPS Panel Full HD 1920x1080  Intel Core i5 8250U 1.6GHz   
10                                   1366x768  Intel Core i5 7200U 2.5GHz   
11                          Full HD 1920x1080    Intel Core i3 6006U 2GHz   
12         IPS Panel Retina Display 2880x1800        Intel Core i7 2.8GHz   
13                          Full HD 1920x1080    Intel Core i3 6006U 2GHz   
14         IPS Panel Retina Display 2304x1440      Intel Core M m3 1.2GHz   
15         IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   
16                          Full HD 1920x1080  Intel Core i7 7500U 2.7GHz   
17         IPS Panel Retina Display 2880x1800        Intel Core i7 2.9GHz   
18                          Full HD 1920x1080  Intel Core i3 7100U 2.4GHz   
19  IPS Panel Full HD / Touchscreen 1920x1080  Intel Core i5 8250U 1.6GHz   

     Ram               Memory                           Gpu       OpSys  \
0    8GB            128GB SSD  Intel Iris Plus Graphics 640       macOS   
1    8GB  128GB Flash Storage        Intel HD Graphics 6000       macOS   
2    8GB            256GB SSD         Intel HD Graphics 620       No OS   
3   16GB            512GB SSD            AMD Radeon Pro 455       macOS   
4    8GB            256GB SSD  Intel Iris Plus Graphics 650       macOS   
5    4GB            500GB HDD                 AMD Radeon R5  Windows 10   
6   16GB  256GB Flash Storage       Intel Iris Pro Graphics    Mac OS X   
7    8GB  256GB Flash Storage        Intel HD Graphics 6000       macOS   
8   16GB            512GB SSD          Nvidia GeForce MX150  Windows 10   
9    8GB            256GB SSD        Intel UHD Graphics 620  Windows 10   
10   4GB            500GB HDD         Intel HD Graphics 620       No OS   
11   4GB            500GB HDD         Intel HD Graphics 520       No OS   
12  16GB            256GB SSD            AMD Radeon Pro 555       macOS   
13   4GB    

In [9]:
laptop['Memory'].value_counts()

Memory
256GB SSD                        412
1TB HDD                          223
500GB HDD                        132
512GB SSD                        118
128GB SSD +  1TB HDD              94
128GB SSD                         76
256GB SSD +  1TB HDD              73
32GB Flash Storage                38
2TB HDD                           16
64GB Flash Storage                15
1TB SSD                           14
512GB SSD +  1TB HDD              14
256GB SSD +  2TB HDD              10
1.0TB Hybrid                       9
256GB Flash Storage                8
16GB Flash Storage                 7
32GB SSD                           6
180GB SSD                          5
128GB Flash Storage                4
16GB SSD                           3
512GB SSD +  2TB HDD               3
128GB SSD +  2TB HDD               2
256GB SSD +  256GB SSD             2
512GB Flash Storage                2
1TB SSD +  1TB HDD                 2
256GB SSD +  500GB HDD             2
64GB SSD                       

## Задание 2.
2.1 Создайте признак SSD, который изначально равен 0
2.2 Поставьте в признаке SSD 1, если ноутбук
действительно с типом носителя SSD
2.3 Уберите в признаке Weight значения 'kg' и поменяйте его
тип данных на вещественный

In [16]:
# Создаем признак SSD изначально равный 0
laptop['SSD'] = 0

# Вариант 1: Простая логика - если есть SSD в любом виде
laptop['SSD'] = laptop['Memory'].str.contains('SSD', case=False, na=False).astype(int)
laptop.head()

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   
3          4   Apple  MacBook Pro  Ultrabook    15.4   
4          5   Apple  MacBook Pro  Ultrabook    13.3   

                     ScreenResolution                         Cpu   Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

                Memory                           Gpu  OpSys  Weight  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS  1.37kg   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS  1.34kg   
2            256GB SSD         Intel HD Graphics 620  No OS  1.86kg   
3            512GB SSD            AMD Radeon Pro 455  macOS  1.83kg   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS  1.37kg   

   Price_euros Cpu_Company Memory_Amount Memory_Type  Memory_Amount1  \
0      1339.69       Intel         128GB         SSD             128   
1       898.94       Intel         128GB       Flash             128   
2       575.00       Intel         256GB         SSD             256   
3      2537.45       Intel         512GB         SSD             512   
4      1803.60       Intel         256GB         SSD             256   

   Memory_Amount2  SSD  SSD1  
0             128    1     0  
1             128    0     0  
2             256    1     0  
3             512    1     0  
4             256    1     0

In [19]:
# ИЛИ Вариант 2: Более точная логика с учетом комбинаций
def check_ssd(memory_text):
    memory_str = str(memory_text).upper()
    if 'SSD' in memory_str:
        return 1
    return 0

laptop['SSD1'] = laptop['Memory'].apply(check_ssd)
laptop.head()

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   
3          4   Apple  MacBook Pro  Ultrabook    15.4   
4          5   Apple  MacBook Pro  Ultrabook    13.3   

                     ScreenResolution                         Cpu   Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

                Memory                           Gpu  OpSys  Weight  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS  1.37kg   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS  1.34kg   
2            256GB SSD         Intel HD Graphics 620  No OS  1.86kg   
3            512GB SSD            AMD Radeon Pro 455  macOS  1.83kg   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS  1.37kg   

   Price_euros Cpu_Company Memory_Amount Memory_Type  Memory_Amount1  \
0      1339.69       Intel         128GB         SSD             128   
1       898.94       Intel         128GB       Flash             128   
2       575.00       Intel         256GB         SSD             256   
3      2537.45       Intel         512GB         SSD             512   
4      1803.60       Intel         256GB         SSD             256   

   Memory_Amount2  SSD  SSD1  
0             128    1     1  
1             128    0     0  
2             256    1     1  
3             512    1     1  
4             256    1     1

In [20]:
# Создадим несколько признаков для разных типов накопителей
laptop['Storage_SSD'] = laptop['Memory'].str.contains('SSD', case=False, na=False).astype(int)
laptop['Storage_HDD'] = laptop['Memory'].str.contains('HDD', case=False, na=False).astype(int)
laptop['Storage_Flash'] = laptop['Memory'].str.contains('Flash', case=False, na=False).astype(int)
laptop['Storage_Hybrid'] = laptop['Memory'].str.contains('Hybrid', case=False, na=False).astype(int)

# Посмотрим распределение
print("SSD:", laptop['Storage_SSD'].sum())
print("HDD:", laptop['Storage_HDD'].sum()) 
print("Flash:", laptop['Storage_Flash'].sum())
print("Hybrid:", laptop['Storage_Hybrid'].sum())

SSD: 843
HDD: 576
Flash: 75
Hybrid: 12


In [21]:
# Количество типов накопителей в одном ноутбуке
laptop['Storage_Types_Count'] = (laptop['Storage_SSD'] + 
                                laptop['Storage_HDD'] + 
                                laptop['Storage_Flash'] + 
                                laptop['Storage_Hybrid'])

# Проверим комбинации
print(laptop['Storage_Types_Count'].value_counts())

Storage_Types_Count
1    1100
2     203
Name: count, dtype: int64


In [23]:
# Посмотрим примеры с разными типами накопителей
sample_data = laptop[['Memory', 'Storage_SSD', 'Storage_HDD', 'Storage_Flash', 'Storage_Hybrid']]
print(sample_data.head())

                Memory  Storage_SSD  Storage_HDD  Storage_Flash  \
0            128GB SSD            1            0              0   
1  128GB Flash Storage            0            0              1   
2            256GB SSD            1            0              0   
3            512GB SSD            1            0              0   
4            256GB SSD            1            0              0   

   Storage_Hybrid  
0               0  
1               0  
2               0  
3               0  
4               0  


In [24]:
import openpyxl
print("openpyxl успешно установлен!")

openpyxl успешно установлен!


In [ ]:
# Самый простой способ
laptop.to_excel('laptop_data.xlsx', index=False)

# С дополнительными параметрами
laptop.to_excel('laptop_data.xlsx', 
                index=False,          # не сохранять индексы
                sheet_name='Laptops') # название листа

In [2]:
#Открытие в Excel после сохранения
import os
import subprocess

# Сохраняем
# laptop.to_excel('laptop_data1.xlsx', index=False)

# Открываем в Excel (Windows)
# os.startfile('laptop_data1.xlsx')  # для Windows

# Или кроссплатформенный способ
subprocess.run(['start', 'laptop_data1.xlsx'], shell=True)  # Windows
# subprocess.run(['open', 'laptop_data.xlsx'])  # Mac
# subprocess.run(['xdg-open', 'laptop_data.xlsx'])  # Linux

CompletedProcess(args=['start', 'laptop_data1.xlsx'], returncode=0)

In [3]:
def convert_to_type(x):
    return x.replace(' ', '').split('B[-1]')

laptop['Memory_Type1'] = laptop['Memory'].apply(convert_to_type)
laptop.head()

NameError: name 'laptop' is not defined

## Задание 3.
Создайте датафрейм с клиентами:
clients = pd.DataFrame({
'client_id': [45, 32, 67, 33, 43],
'laptop_id': [506, 398, 710, 120, 1999]
})
laptop_id - это индексы датафрейма с ноутбуками
3.1 Присоедините к таблице clients данные по ноутбукам
через метод join
3.2 Присоедините к таблице clients данные по ноутбукам
через метод merge
Это нужно, чтобы понимать, какие ноутбуки покупались клиентами

## Задание 4.
Составьте несколько сводных таблиц
4.1 Найдите среднюю стоимость ноутбуков в зависимости
от компании производителя
Отсортируйте от меньшей стоимости к большей
4.2 Найдите минимальную, среднюю и максимальную
стоимости ноутбуков в зависимости от производителя
процессора
4.3 Постройте таблицу с подсчетом количества ноутбуков в
данных в зависимости от производителя CPU и ОЗУ
4.4 Постройте таблицу с подсчетом средней стоимости
ноутбуков в данных в зависимости от операционной
системы и GB памяти

## Задание 5*.
Ответьте на несколько вопросов
5.1 Ноутбуков каких компаний и с каким процессором
больше?
5.2 С каким типом памяти и с каким объемом памяти
больше ноутбуков?


## Домашнее задание 1
Изменение таблиц в Pandas
Загрузите данные из файла kc-house-data.csv
Считать данные с помощью pandas
Вывести на экран первые 5 строк
1.1 Создать новый признак price_per_sq_lot,
который будет содержать среднюю стоимость за
один кв. метр общей площади
1.2 Создать новый признак delta_renovated,
который будет содержать разницу в годах между
годом реновацией дома и годом постройки дома
Если реновации дома не было, то в новом признаке поставьте 0

## Домашнее задание 2
Изменение таблиц в Pandas
Создайте датафрейм с клиентами:
clients = pd.DataFrame({
'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295, 939, 3840, 280, 20, 4332, 3475, 4213, 3113,
4809, 2134, 2242, 2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060, 1189000207, 6821600300, 7137950720,
9510920050, 6131600255, 5428000070, 1788800910, 8100400160, 3123049142, 6306800010, 5083000375, 7920100025,
1951600150, 809001400, 339600110, 1622049154, 1099600250, 8563000110, 2768100205, 3995700435, 8861700030,
3303980210, 7731100066, 8146100580, 825069097, 3889100029, 9524100196]
})
house_id - это индексы датафрейма с домами
2.1 Присоедините к таблице clients данные по домам через метод join
2.2 Присоедините к таблице clients данные по домам через метод merge

## Домашнее задание 3
Изменение таблиц в Pandas
Составьте несколько сводных таблиц
3.1 Найдите среднюю стоимость домов в
зависимости от количества спален
Отсортируйте от меньшей стоимости к большей
3.2 Найдите минимальную, среднюю и
максимальную стоимости домов в зависимости
от состояния дома
3.3 Постройте таблицу с подсчетом количества
домов в данных в зависимости от вида на
набережную и оценкой вида